### PR TITLE
#11535 - Add a create/edit form for environments in the web app

### LINKED ISSUE #11535: Add a create/edit form for environments in the web app [5]
### Feature Description

After an Environment entity has been created in https://github.com/SORMAS-Foundation/SORMAS-Project/issues/11796 as part of epic https://github.com/SORMAS-Foundation/SORMAS-Project/issues/11533, create and edit forms for this entity are now required for the web app.

### Added Value/Benefit
Users will be able to create and edit environment entities for further usage in the system.

### Acceptance Criteria

- [x] A creation form for environments has been added with the following fields:
  - Report date (required)
  - External ID
  - Environment name  (required)
  - Environment media (required)
  - Location form (region, district, gps lat and gps lon required; region and district pre-filled with the jurisdiction of the user if available)
  - Responsible user (contains a list of all users with the `ENVIRONMENT_EDIT` user right that are assigned to the district selected in the location form)
- [x] An edit form for environments has been added with all fields from the creation form and the following additional fields:
  - UUID (read only)
  - Reporting user (read only)
  - Investigation status (required)
  - Water type (if environment media == water)
  - Other water type (if water type == other)
  - Infrastructure details (if environment media == water)
  - Other infrastructure details (if infrastructure details == other)
  - A "Water use" heading (if environment media == water)
  - Checkboxes for all `WaterUse` enum values (if environment media == water); see the epidemiologicalEvidenceDetails checkboxes of events for reference
  - Other water use (if the checkbox for "Other" water use has been checked)
- [x] A "New Environment" button has been added to the environment directory created in #11534 that brings up the creation form
- [x] A column with an edit button has been added to the environment directory created in #11534 that takes the user to a dedicated page for the environment, displaying the edit form
- [x] The edit form contains functional buttons to delete and archive the environment if the user has the respective user rights
- [x] The location form includes the functionality to automatically pick the GPS coordinates based on the address details entered (already implemented; just verify that it's working)
- [x] The unsaved changes warning appears as expected for the edit form
- [x] The top navigation when inside the environment edit page contains a backlink that takes the user back to the environment directory

### Mockups
* creation form
image

* edit form
env_editform

### Additional Information

The tasks and samples on the right are not supposed to be implemented in this issue.

### PR DESCRIPTION
**Pull request by sergiupacurariu**
Fixes #11535

### CODE REVIEW

File: sormas-api/src/main/java/de/symeda/sormas/api/environment/EnvironmentFacade.java [hunk 1]
```diff
@@ -11,4 +11,6 @@
 public interface EnvironmentFacade extends CoreFacade<EnvironmentDto, EnvironmentIndexDto, EnvironmentReferenceDto, EnvironmentCriteria> {
 
 	List<EnvironmentDto> getAllEnvironmentsAfter(Date date);
+
+	EnvironmentDto getEnvironmentDataByUuid(String uuid);
```
Comment (leventegal-she) [inline sormas-api/src/main/java/de/symeda/sormas/api/environment/EnvironmentFacade.java:15]: this method exists. please use `getByUuid` of `AbstractBaseEjb`

**Comment by sormas-vitagroup** [comment]
SonarCloud analysis: https://sonarcloud.io/dashboard?id=SORMAS-Project&pullRequest=12240

File: sormas-ui/src/main/java/de/symeda/sormas/ui/environment/EnvironmentController.java [hunk 2]
```diff
@@ -15,11 +15,162 @@
...
+	}
+
+	public CommitDiscardWrapperComponent<EnvironmentCreateForm> getEnvironmentCreateComponent() {
+		UserProvider curentUser = UserProvider.getCurrent();
+
+		if (curentUser != null) {
+			EnvironmentCreateForm createForm;
+			createForm = new EnvironmentCreateForm();
+			final EnvironmentDto environment = EnvironmentDto.build();
+			environment.setReportingUser(curentUser.getUserReference());
```
Comment (leventegal-she) [inline sormas-ui/src/main/java/de/symeda/sormas/ui/environment/EnvironmentController.java:63]: please add a `build` method in the `EnvironmentDto` that takes a user and sets these values on the new environment

File: sormas-ui/src/main/java/de/symeda/sormas/ui/environment/EnvironmentController.java [hunk 3]
```diff
@@ -15,11 +15,162 @@
...
+			environment.getLocation().setRegion(curentUser.getUser().getRegion());
+			environment.getLocation().setDistrict(curentUser.getUser().getDistrict());
+			environment.setInvestigationStatus(InvestigationStatus.PENDING);
```
Comment (leventegal-she) [inline sormas-ui/src/main/java/de/symeda/sormas/ui/environment/EnvironmentController.java:66]: this should be in the base `build()` method

File: sormas-ui/src/main/java/de/symeda/sormas/ui/environment/EnvironmentCreateForm.java [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+			} else {
+				districtEnvironmentResponsibles.clear();
+			}
+			addDistrict(responsibleUserField);
+		});
+
+		setRequired(true, EnvironmentDto.REPORT_DATE, EnvironmentDto.ENVIRONMENT_MEDIA, EnvironmentDto.ENVIRONMENT_NAME);
+	}
+
+	private void addDistrict(ComboBox responsibleUserField) {
```
Comment (leventegal-she) [inline sormas-ui/src/main/java/de/symeda/sormas/ui/environment/EnvironmentCreateForm.java:90]: please rename to `updateResponsibleUsersCombo` to explain what it does

also this method could take the user list as an argument so there would be no need to store it in a class variable

File: sormas-ui/src/main/java/de/symeda/sormas/ui/environment/EnvironmentDataForm.java [hunk 5]
```diff
@@ -0,0 +1,229 @@
...
+
+	private boolean isWaterUseOtherChecked() {
+		return Boolean.TRUE.equals(waterUseCheckBoxTree.getValues().get(WaterUse.OTHER));
+	}
+
+	private CheckBoxTree.CheckBoxElement<WaterUse> environmentEvidenceDetailToCheckBoxElement(WaterUse waterUse) {
+		return new CheckBoxTree.CheckBoxElement<>(null, waterUse);
+	}
+
+	private void addDistrict(ComboBox responsibleUserField) {
```
Comment (leventegal-she) [inline sormas-ui/src/main/java/de/symeda/sormas/ui/environment/EnvironmentDataForm.java:207]: same as in the create form

File: sormas-backend/src/main/java/de/symeda/sormas/backend/environment/EnvironmentFacadeEjb.java [hunk 6]
```diff
@@ -301,6 +314,36 @@ public static EnvironmentReferenceDto toReferenceDto(Environment entity) {
...
+	}
+
+	@Override
+	@RightsAllowed(UserRight._ENVIRONMENT_DELETE)
+	public void restore(String uuid) {
+		super.restore(uuid);
+	}
+
 	@Override
 	protected void pseudonymizeDto(Environment source, EnvironmentDto dto, Pseudonymizer pseudonymizer, boolean inJurisdiction) {
```
Comment (leventegal-she) [inline sormas-backend/src/main/java/de/symeda/sormas/backend/environment/EnvironmentFacadeEjb.java:348]: `pseudonymizeDto` and `restorePseudonymizedDto` should be implemented

**Review by leventegal-she — CHANGES_REQUESTED** [review]

**Comment by sormas-vitagroup** [comment]
SonarCloud analysis: https://sonarcloud.io/dashboard?id=SORMAS-Project&pullRequest=12240

File: sormas-api/src/main/java/de/symeda/sormas/api/environment/EnvironmentFacade.java (hunk 1 above)
Comment (sergiupacurariu) [reply sormas-api/src/main/java/de/symeda/sormas/api/environment/EnvironmentFacade.java:15]: removed

File: sormas-ui/src/main/java/de/symeda/sormas/ui/environment/EnvironmentController.java (hunk 2 above)
Comment (sergiupacurariu) [reply sormas-ui/src/main/java/de/symeda/sormas/ui/environment/EnvironmentController.java:63]: updated

File: sormas-ui/src/main/java/de/symeda/sormas/ui/environment/EnvironmentController.java (hunk 3 above)
Comment (sergiupacurariu) [reply sormas-ui/src/main/java/de/symeda/sormas/ui/environment/EnvironmentController.java:66]: updated

File: sormas-ui/src/main/java/de/symeda/sormas/ui/environment/EnvironmentCreateForm.java (hunk 4 above)
Comment (sergiupacurariu) [reply sormas-ui/src/main/java/de/symeda/sormas/ui/environment/EnvironmentCreateForm.java:90]: removed the method

File: sormas-ui/src/main/java/de/symeda/sormas/ui/environment/EnvironmentDataForm.java (hunk 5 above)
Comment (sergiupacurariu) [reply sormas-ui/src/main/java/de/symeda/sormas/ui/environment/EnvironmentDataForm.java:207]: removed the method

File: sormas-backend/src/main/java/de/symeda/sormas/backend/environment/EnvironmentFacadeEjb.java (hunk 6 above)
Comment (sergiupacurariu) [reply sormas-backend/src/main/java/de/symeda/sormas/backend/environment/EnvironmentFacadeEjb.java:348]: implemented

**Comment by leventegal-she** [comment]
Environment tab should not redirect to the environment list
image

Discard doesn't work on the *Use of water*  checkboxes

**Comment by sergiupacurariu** [comment]
> Environment tab should not redirect to the environment list image
> 
> Discard doesn't work on the _Use of water_ checkboxes

the environment link issue was fixed
Discard issue is the same both in CaseData and EventData. I would create a new ticket for all three.

**Comment by sormas-vitagroup** [comment]
SonarCloud analysis: https://sonarcloud.io/dashboard?id=SORMAS-Project&pullRequest=12240

**Comment by sormas-vitagroup** [comment]
SonarCloud analysis: https://sonarcloud.io/dashboard?id=SORMAS-Project&pullRequest=12240

**Comment by sormas-vitagroup** [comment]
SonarCloud analysis: https://sonarcloud.io/dashboard?id=SORMAS-Project&pullRequest=12240

File: sormas-ui/src/main/java/de/symeda/sormas/ui/environment/EnvironmentCreateForm.java [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+
+		ComboBox responsibleUserField = addField(EnvironmentDto.RESPONSIBLE_USER, ComboBox.class);
+		responsibleUserField.setNullSelectionAllowed(true);
+
+		districtField.addValueChangeListener(e -> {
+			DistrictReferenceDto district = (DistrictReferenceDto) districtField.getValue();
+			List<UserReferenceDto> districtEnvironmentResponsibles = new ArrayList<>();
+			if (district != null) {
+				districtEnvironmentResponsibles = FacadeProvider.getUserFacade().getUserRefsByDistrict(district, true, UserRight.ENVIRONMENT_EDIT);
+			} else {
```
Comment (leventegal-she) [inline sormas-ui/src/main/java/de/symeda/sormas/ui/environment/EnvironmentCreateForm.java:85]: the else block is not needed

File: sormas-ui/src/main/java/de/symeda/sormas/ui/environment/EnvironmentDataForm.java [hunk 8]
```diff
@@ -0,0 +1,220 @@
...
+		addField(EnvironmentDto.OTHER_DELETION_REASON, TextArea.class).setRows(3);
+		setVisible(false, EnvironmentDto.DELETION_REASON, EnvironmentDto.OTHER_DELETION_REASON);
+
+		districtField.addValueChangeListener(e -> {
+			DistrictReferenceDto district = (DistrictReferenceDto) districtField.getValue();
+			List<UserReferenceDto> districtEnvironmentResponsibles = new ArrayList<>();
+			if (district != null) {
+				districtEnvironmentResponsibles = FacadeProvider.getUserFacade().getUserRefsByDistrict(district, true, UserRight.ENVIRONMENT_EDIT);
+			} else {
+				districtEnvironmentResponsibles.clear();
```
Comment (leventegal-she) [inline sormas-ui/src/main/java/de/symeda/sormas/ui/environment/EnvironmentDataForm.java:159]: the else block is not needed

**Review by leventegal-she — CHANGES_REQUESTED** [review]
The address fields are not pseudonymized on the backend and they are editable in the webapp

File: sormas-api/src/main/java/de/symeda/sormas/api/environment/EnvironmentDto.java [hunk 9]
```diff
@@ -59,9 +66,25 @@ public class EnvironmentDto extends EntityDto {
...
+
+		return environment;
+	}
+
+	public static EnvironmentDto build(UserDto currentUser) {
+		EnvironmentDto environment = build();
+		environment.setReportingUser(currentUser.toReference());
+		environment.getLocation().setRegion(currentUser.getRegion());
+		environment.getLocation().setDistrict(currentUser.getDistrict());
+		environment.setInvestigationStatus(InvestigationStatus.PENDING);
```
[the remark this answers is not in the document]
Comment (sergiupacurariu) [reply sormas-api/src/main/java/de/symeda/sormas/api/environment/EnvironmentDto.java:87]: updated

File: sormas-ui/src/main/java/de/symeda/sormas/ui/environment/EnvironmentCreateForm.java (hunk 7 above)
Comment (sergiupacurariu) [reply sormas-ui/src/main/java/de/symeda/sormas/ui/environment/EnvironmentCreateForm.java:85]: removed

File: sormas-ui/src/main/java/de/symeda/sormas/ui/environment/EnvironmentDataForm.java (hunk 8 above)
Comment (sergiupacurariu) [reply sormas-ui/src/main/java/de/symeda/sormas/ui/environment/EnvironmentDataForm.java:159]: removed

**Comment by sormas-vitagroup** [comment]
SonarCloud analysis: https://sonarcloud.io/dashboard?id=SORMAS-Project&pullRequest=12240

**Review by leventegal-she — APPROVED** [review]

**Pull request merged**